pilot/pkg/serviceregistry/kube/controller/nodeuntainter.go

```diff
@@ -0,0 +1,153 @@
+	queue       controllers.Queue
+}
+
+func newNodeUntainter(c *Controller) *nodeUntainter {
+	revision := c.opts.Revision
+	if revision == "" {
+		revision = "default"
+	}
+	labels := map[string]string{
+		"k8s-app":      "istio-cni-node",
+		"istio.io/rev": revision,
+	}
+	log.Debugf("starting node untainter with labels %v", labels)
+	nt := &nodeUntainter{
+		podsClient:  c.podsClient,
+		nodesClient: c.nodes,
+		cnilabels:   labels,
+		// TODO: need to support the cni deployed to kube-system
+		// need to get somehow the value of components.cni.namespace to here.
+		ourNs: c.opts.CniNamespace,
+	}
+	nt.setup()
+	return nt
+}
+
+func (n *nodeUntainter) setup() {
+	nodes := krt.WrapClient[*v1.Node](n.nodesClient)
+	// as we fetch all pods anyway, no need to create new informer..
+	pods := krt.WrapClient[*v1.Pod](n.podsClient)
+
+	labelInst := labels.Instance(n.cnilabels)
+	readyCniPods := krt.NewCollection(pods, func(ctx krt.HandlerContext, p *v1.Pod) *v1.Pod {
+		log.Debugf("cniPods event: %s", p.Name)
+		if p.Namespace != n.ourNs {
+			return nil
+		}
+		if !labelInst.SubsetOf(p.ObjectMeta.Labels) {
+			return nil
+		}
+		if !IsPodReadyConditionTrue(p.Status) {
+			return nil
+		}
+		log.Debugf("pod %s on node %s ready!", p.Name, p.Spec.NodeName)
+		return p
+	})
+
+	// these are all the nodes that have a ready cni pod. if the cni pod is ready,
+	// it means we are ok scheduling pods to it.
+	readyCniNodes := krt.NewCollection(readyCniPods, func(ctx krt.HandlerContext, p v1.Pod) *v1.Node {
+		pnode := krt.FetchOne(ctx, nodes, krt.FilterKey(p.Spec.NodeName))
+		if pnode == nil {
+			return nil
+		}
+		node := *pnode
+		if !hasTaint(node) {
+			return nil
+		}
+		return node
+	})
+
+	n.queue = controllers.NewQueue("untaint nodes",
+		controllers.WithReconciler(n.reconcileNode),
+		controllers.WithMaxAttempts(5))
+
+	// remove the taints from readyCniNodes
+	readyCniNodes.Register(func(o krt.Event[v1.Node]) {
+		if o.Event == controllers.EventDelete {
+			return
+		}
+		if o.New != nil {
+			log.Debugf("adding node to queue event: %s", o.New.Name)
+			n.queue.AddObject(o.New)
+		}
+	})
+}
+
+func (n *nodeUntainter) Run(stop <-chan struct{}) {
+	n.queue.Run(stop)
+}
+
+func (n *nodeUntainter) reconcileNode(key types.NamespacedName) error {
+	log.Debugf("reconciling node %s", key.Name)
+	node := n.nodesClient.Get(key.Name, key.Namespace)
+	if node == nil {
+		return nil
+	}
+
+	err := removeReadinessTaint(n.nodesClient, node)
+	if err != nil {
+		log.Errorf("failed to remove readiness taint from node %v: %v", node.Name, err)
+	}
+	return err
+}
+
+func removeReadinessTaint(nodesClient kclient.Client[*v1.Node], node *v1.Node) error {
+	updatedTaint := deleteTaint(node.Spec.Taints, &v1.Taint{Key: TaintName, Effect: v1.TaintEffectNoSchedule})
+	if len(updatedTaint) == len(node.Spec.Taints) {
+		return nil
+	}
+
+	log.Debugf("removing readiness taint from node %v", node.Name)
+	node.Spec.Taints = updatedTaint
+	_, err := nodesClient.Update(node)
```

Comment:
nit: does a patch work? then we don't have to worry about conflicts. Not sure how easy it is to do remove with patch though

Reply:
patch might work but the client doesn't have a patch method :/

Reply:
I added patch in another PR fwiw